dotnet/src/Plugins/Plugins.Core/CodeInterpreter/SessionsPythonPlugin.cs

```diff
@@ -0,0 +1,304 @@
+
+        var response = await httpClient.GetAsync($"{this._poolManagementEndpoint}python/downloadFile?identifier={this._settings.SessionId}&filename={remoteFilePath}").ConfigureAwait(false);
+        if (!response.IsSuccessStatusCode)
+        {
+            var errorBody = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            throw new HttpRequestException($"Failed to download file. Status code: {response.StatusCode}. Details: {errorBody}");
+        }
+
+        var fileContent = await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+
+        if (!string.IsNullOrWhiteSpace(localFilePath))
+        {
+            try
+            {
+                File.WriteAllBytes(localFilePath, fileContent);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException("Failed to write file to disk.", ex);
+            }
+        }
+
+        return fileContent;
+    }
+
+    /// <summary>
+    /// Lists all files in the provided session id pool.
+    /// </summary>
+    /// <returns> The list of files in the session. </returns>
+    [KernelFunction, Description("Lists all files in the provided session id pool")]
+    public async Task<IReadOnlyList<SessionRemoteFileMetadata>> ListFilesAsync()
+    {
+        if (this._logger.IsEnabled(LogLevel.Trace))
+        {
+            this._logger.LogTrace("Listing files for Session ID: {SessionId}", this._settings.SessionId);
+        }
+
+        using var httpClient = this._httpClientFactory.CreateClient();
+        httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {(await this.GetAuthTokenAsync().ConfigureAwait(false))}");
+        httpClient.DefaultRequestHeaders.Add("User-Agent", $"{HttpHeaderConstant.Values.UserAgent}/{s_assemblyVersion} (Language=dotnet)");
+
+        var response = await httpClient.GetAsync($"{this._poolManagementEndpoint}python/files?identifier={this._settings.SessionId}").ConfigureAwait(false);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            throw new HttpRequestException($"Failed to list files. Status code: {response.StatusCode}");
+        }
+
+        var jsonElementResult = JsonSerializer.Deserialize<JsonElement>(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+
+        var files = jsonElementResult.GetProperty("$values");
+
+        var result = new SessionRemoteFileMetadata[files.GetArrayLength()];
+
+        for (var i = 0; i < result.Length; i++)
+        {
+            result[i] = JsonSerializer.Deserialize<SessionRemoteFileMetadata>(files[i].GetRawText())!;
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    /// Retrieves the auth token for the session.
+    /// </summary>
+    private async Task<string> GetAuthTokenAsync()
+    {
+        if (string.IsNullOrWhiteSpace(this._authToken))
```

Comment:
Consider leaving token caching to the authTokenProvider, since they will know best how long the session will last, and whether the token may need to be renewed at some stage.